### PR TITLE
backend: add retry library

### LINKED ISSUE #790: backend: observable retry library
**Description**

Build a generic retry library, with a redeeming quality of being observable such as logging and stats. The motivation behind this issue is that client libraries such as the `aws-go-sdk` have retry-able functionality built in however its not possible observe these retries or why the client library is even retrying at all.

This retry library should implement all of the basic retry / backoff / jitter techniques, in a as enough of a generic way to use this throughout the codebase.

Complexity [S/M/L]: S

### PR DESCRIPTION
**Pull request by goku321**
### Description

backend: add retry library

### Testing Performed

Added test cases

### GitHub Issue

Fixes #790

### CODE REVIEW

**Comment by dschaller** [comment]
Thanks for the contribution @goku321! The team will review this week.

File: backend/retry/options.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+// ExponentialBackoff returns exponentially increasing backoffs by a power of 2.
+func ExponentialBackoff(i uint, _ *Config) time.Duration {
+	return time.Duration(1<<i) * time.Second
+}
+
+// FixedBackoff returns a fixed delay duration.
+func FixedBackoff(_ uint, c *Config) time.Duration {
+	return c.maxDelay
+}
```
Comment (danielhochman) [inline backend/retry/options.go:70]: this is identical to DefaultBackoff

File: backend/retry/options.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+// DefaultBackoff always returns 1 second.
+func DefaultBackoff(_ uint, c *Config) time.Duration {
+	return c.maxDelay
+}
+
+// ExponentialBackoff returns exponentially increasing backoffs by a power of 2.
+func ExponentialBackoff(i uint, _ *Config) time.Duration {
+	return time.Duration(1<<i) * time.Second
```
Comment (danielhochman) [inline backend/retry/options.go:64]: time.Duration is technically converting to nanos which is not necessary
```suggestion
	return 1<<i * time.Second
```

Comment (danielhochman) [reply backend/retry/options.go:64]: also can we shorten the initial backoff to 100ms or similar? starting with 1 second feels very long, but everyone's systems are different i suppose

File: backend/retry/retry.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+			// Avoid waiting if this is the last attempt.
+			if n == config.maxRetries - 1 {
+				break
+			}
+
+			delayTime := config.backoff(n, config)
+
+			select {
+			case <-time.After(delayTime):
```
Comment (danielhochman) [inline backend/retry/retry.go:74]: nit: comment empty case for clarity
```suggestion
			case <-time.After(delayTime):
				// Time for next retry.
```

**Review by danielhochman — COMMENTED** [review]
very nice and useful! some minor comments, plus you will need to fix lint.

**Comment by goku321** [comment]
@danielhochman Thank you for reviewing the PR. I've made the required changes 🙂

**Review by danielhochman — APPROVED** [review]
looks great to me! will need a second reviewer now, i'll ping someone.

File: backend/retry/retry.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+// RetryableFunc represents a function that is retryable.
+type RetryableFunc func() error
+
+// MultiError is a slice of errors ocurred while retrying.
+type MultiError []error
+
+// Error implements error interface.
+//
+// Returns all the errors ocurred during retry formatted as a string.
```
Comment (dschaller) [inline backend/retry/retry.go:21]: nit:
```suggestion
// Returns all the errors occurred during retry formatted as a string separated by newlines.
```

File: backend/retry/retry.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	errorLog := make([]string, len(m))
+	for i, e := range m {
+		errorLog[i] = fmt.Sprintf("Retry #%d: %s", i+1, e.Error())
+	}
+
+	return fmt.Sprintf("All retries failed:\n%s", strings.Join(errorLog, "\n"))
+}
+
+/*
+Do retries a given function with delays between each attempts as defined by the caller.
```
Comment (dschaller) [inline backend/retry/retry.go:32]: nit:
```suggestion
Do retries a given function with delays between each attempts as defined by the caller or their corresponding defaults.
```

**Review by dschaller — APPROVED** [review]
Looks great! A few minor comments for docs and then we can merge

File: backend/retry/retry.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	n := uint(0)
+	var errorLog MultiError
+
+	config := defaultConfig(scope)
+	// apply user defined options.
+	for _, opt := range opts {
+		opt(config)
+	}
+
+	for n < config.maxRetries {
```
Comment (dschaller) [inline backend/retry/retry.go:57]: I think you can simplify some of the logic by inlining n here
```go
for n := 1; n < config.maxRetries; n++ {
```

**Review by dschaller — APPROVED** [review]

**Comment by dschaller** [comment]
Thanks @goku321!

**Pull request merged**